File: SERVICE_microscopic_simulation/src/player_database.js
const fs = require('fs');
const dgram = require('dgram');
const server = dgram.createSocket('udp4');
const playerStateSize = 100;

let playerDataMap = new Map();

server.on('message', (msg, rinfo) => {
  let playerId = parseInt(msg.slice(0, 4).toString());
  const offset = playerId * playerStateSize;
  playerDataMap.set(rinfo, {playerId: playerId, offset: offset, data: msg});
  let buffer = Buffer.from(msg);
  fs.writeFile(`player_${playerId}.bin`, buffer, 'binary', (err) => {
    if(err) {
      console.log(err);
    } else {
      console.log(`The file was saved as player_${playerId}.bin!`);
    }
  });
});

server.bind(8000);

File: SERVICE_microscopic_simulation/sample_project/player_database.exs
# Copyright (c) 2018-present. This file is part of V-Sekai https://v-sekai.org/.
# K. S. Ernest (Fire) Lee & Contributors
# player_database.exs
# SPDX-License-Identifier: MIT

defmodule PlayerDatabase do
  def start do
    {:ok, socket} = :gen_udp.open(8000, [:binary, active: false])
    loop(socket)
  end

  defp loop(socket) do
    {:ok, {ip, port, msg}} = :gen_udp.recv(socket, 0)
    player_id = String.slice(msg, 0..3) |> String.to_integer()
    offset = player_id * 100
    player_data_map = %{ip: ip, port: port, player_id: player_id, offset: offset, data: msg}
    File.write("player_#{player_id}.bin", msg)
    IO.puts "The file was saved as player_#{player_id}.bin!"
    loop(socket)
  end
end

PlayerDatabase.start()
